binding/dotnet/RhinoTest/MainTest.cs

```diff
@@ -35,96 +36,236 @@ public static void ClassInitialize(TestContext testContext)
+        private static string GetModelPath(string language)
         {
-            using Rhino r = SetUpClass();
-            Assert.IsFalse(string.IsNullOrWhiteSpace(r?.ContextInfo), "Rhino did not return any context information.");
+            string file_name = AppendLanguage("rhino_params", language);
+            return Path.Combine(
+                _relativeDir,
+                "../../../../../../lib/common",
+                $"{file_name}.pv"
+            );
         }
 
-        [TestMethod]
-        public void TestWithinContext()
+        private void RunTestCase(string audioFileName, bool isWithinContext, string expectedIntent = null, Dictionary<string, string> expectedSlots = null)
         {
-            using Rhino r = SetUpClass();
-            int frameLen = r.FrameLength;
-
-            string testAudioPath = Path.Combine(_relativeDir, "resources/audio_samples/test_within_context.wav");
-            List<short> data = GetPcmFromFile(testAudioPath, r.SampleRate);
+            int frameLen = rhino.FrameLength;
+            string testAudioPath = Path.Combine(_relativeDir, "resources/audio_samples", audioFileName);
+            List<short> data = GetPcmFromFile(testAudioPath, rhino.SampleRate);
 
             bool isFinalized = false;
             int framecount = (int)Math.Floor((float)(data.Count / frameLen));
             var results = new List<int>();
             for (int i = 0; i < framecount; i++)
             {
-                int start = i * r.FrameLength;
-                int count = r.FrameLength;
+                int start = i * rhino.FrameLength;
+                int count = rhino.FrameLength;
                 List<short> frame = data.GetRange(start, count);
-                isFinalized = r.Process(frame.ToArray());
+                isFinalized = rhino.Process(frame.ToArray());
                 if (isFinalized)
                 {
                     break;
                 }
             }
             Assert.IsTrue(isFinalized, "Failed to finalize.");
 
-            Inference inference = r.GetInference();
-            Assert.IsTrue(inference.IsUnderstood, "Couldn't understand.");
-            Assert.AreEqual("orderBeverage", inference.Intent, "Incorrect intent.");
+            Inference inference = rhino.GetInference();
+
+            if(isWithinContext)
+            {
+                Assert.IsTrue(inference.IsUnderstood, "Couldn't understand.");
+                Assert.AreEqual(expectedIntent, inference.Intent, "Incorrect intent.");
+                Assert.IsTrue(inference.Slots.All((keyValuePair) =>
+                                            expectedSlots.ContainsKey(keyValuePair.Key) &&
+                                            expectedSlots[keyValuePair.Key] == keyValuePair.Value));   
+            }
+            else
+            {
+                Assert.IsFalse(inference.IsUnderstood, "Shouldn't be able to understand.");
+            }
+
+            rhino.Dispose();
```

Comment:
I think disposing here introduces a bit of a side-effect problem. When you have a function like `RunTestCase` you really just want it to run a test case for you. If it cleans up some of your resources before it returns, that could be a surprise for some folks who are designing tests in this file in the future. I would rather see a `rhino.Dispose()` in all the test methods than have another function clean it up as a side-effect. 

However, in .NET, we use the IDisposable interface, so if you declare your rhino with `using rhino = ...`, then
`rhino.Dispose()` will be called as soon as you leave the scope of the method!